cpp/arcticdb/storage/memory/memory_storage-inl.hpp

```diff
@@ -55,7 +56,10 @@ namespace arcticdb::storage::memory {
             for (auto &kv : group.values()) {
                 auto it = key_vec.find(kv.variant_key());
 
-                util::check_rte(opts.upsert_ || it != key_vec.end(), "update called with upsert=false but key does not exist");
+                if (!opts.upsert_ && it == key_vec.end()) {
+                    throw KeyNotFoundException(std::move(kv.variant_key()),
+                            "update called with upsert=false but key does not exist: {}");
```

Comment:
The error message is different than the error [message for LMDB](https://github.com/man-group/ArcticDB/pull/584/files#r1283381737). I think we should unify them.

(Minor) I wouldn't move `kv.variant_key()` but copy it. We're already doing the slow thing throwing, so I don't see any benefit of moving a string. If we don't move the debugging will be slightly better as the key will exist in the stack trace.